Woodoku/tests/Entity/test_woodoku_board.py

```diff
@@ -0,0 +1,121 @@
+from typing import Tuple
+
+import pytest
+
+from Exceptions.Exceptions import ShapeOutOfBoardError
+from WoodokuBoard import _WoodokuBoardRepresentation
+import numpy as np
+
+N = 9
+
+
+class TestWoodokuRepresentation:
```

Comment:
For all the test cases here, I would say it is better to not using the underlining `__board` object. Instead, just test with the method calls `add_blocks`, `remove_blocks`, `is_occupied` and `is_not_occupied`. 

For example, `test_add_blocks_add_one_block` is testing if a block is actually added after calling `add_blocks`. This could be written as 

```python
    def test_add_blocks_add_one_block(self):
        board = _WoodokuBoardRepresentation()
        board.add_blocks(self.one_block_lst)
        assert board.is_occupied(self.one_block_lst)
```

You might claim that in this above way, we are not testing `add_blocks` method alone. In fact, we do not need to. The only reason we call `add_blocks` is because we hope to get a `true` value when we call `is_occupied` later. Thus, as long as these two methods have consistent result, it should be acceptable. 

Same logic applies to `remove_blocks` and `is_not_occupied`.

I might also need to borrow some input from @Xiaorui-Huang . What do you think?

Feel free to initiate a discussion!

Reply:
Since I only expose a getter instead of a direct setter,  I thought that it won't cause a large problem, but I will test methods in representation pairwise if both of you agree.

Reply:
> 
> ```python
>     def test_add_blocks_add_one_block(self):
>         board = _WoodokuBoardRepresentation()
>         board.add_blocks(self.one_block_lst)
>         assert board.is_occupied(self.one_block_lst)
> ```
> 
> You might claim that in this above way, we are not testing `add_blocks` method alone. In fact, we do not need to. The only reason we call `add_blocks` is because we hope to get a `true` value when we call `is_occupied` later. Thus, as long as these two methods have consistent result, it should be acceptable.
> 
> Same logic applies to `remove_blocks` and `is_not_occupied`.
Similarly with this example, I might combine two methods and test their combination, eg: `add_block` + `is_occupied`/`is_not_occupied` and `remove_block` + 'is_occupied`/'is_not_occupied'

Reply:
Yeah, I agree using one's own method when testing might not be a great idea. However, the method `board` was created just for testing. It might break the encapsulation principle so other class will also be able to access the inner data object. 

I am proposing to just use the private variable `__board` in test cases, so we could prevent that other non-testing class would have access to it. Thoughts? @YANLING-H @Xiaorui-Huang